src/libutil/util.hh

```diff
@@ -655,5 +655,14 @@ template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
 
 std::string showBytes(uint64_t bytes);
 
+template<typename R = void, typename T = char> inline R withBuffer(size_t size, std::function<R (T[])> fun) {
+    if (size < 0x10000) {
+        T buf[size];
+        return fun(buf);
+    } else {
+        auto buf = std::unique_ptr<T[]>(new T[size]);
+        return fun(buf.get());
+    }
+}
```

Comment:
we can get away with not allocating a `function` here, but keep the signature checking:

```suggestion
template<typename T = char, typename Fn>
inline auto withBuffer(size_t size, Fn fun)
  -> std::invoke_result_t<Fn (T *)>
{
    if (size < 0x10000) {
        T buf[size];
        return fun(buf);
    } else {
        auto buf = std::unique_ptr<T[]>(new T[size]);
        return fun(buf.get());
    }
}
```

`T[]` parameters aren't all that useful usually since they'll immediately decay to pointers everywhere, and they have no range information either